### PR TITLE
Flax Regnet

### PR DESCRIPTION
**Pull request by Shubhamai**
# What does this PR do?

Flax Implementation of [facebook/regnet-y-040](https://huggingface.co/facebook/regnet-y-040)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- Flax: sanchit-gandhi

### CODE REVIEW

**Comment by Shubhamai** [comment]
@sanchit-gandhi This is now ready for your review, thanks a lot for your time.

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 1]
```diff
@@ -0,0 +1,818 @@
...
+
+    ```python
+    >>> from transformers import AutoImageProcessor, FlaxRegNetModel
+    >>> from PIL import Image
+    >>> import requests
+
+    >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    >>> image = Image.open(requests.get(url, stream=True).raw)
+
+    >>> image_processor = AutoImageProcessor.from_pretrained("facebook/regnet-y-040")
```
Comment (Shubhamai) [inline src/transformers/models/regnet/modeling_flax_regnet.py:704]: Made a PR https://huggingface.co/facebook/regnet-y-040/discussions/2

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 2]
```diff
@@ -0,0 +1,818 @@
...
+        self.conv_2 = nn.Conv(
+            self.in_channels,
+            kernel_size=(1, 1),
+            kernel_init=nn.initializers.variance_scaling(2.0, mode="fan_out", distribution="truncated_normal"),
+            dtype=self.dtype,
+            name="2",
+        )  # 2 is the name used in corresponding pytorch implementation
+
+    def __call__(self, hidden_state: jnp.ndarray) -> jnp.ndarray:
+        # b h w c -> b 1 1 c
```
Comment (sanchit-gandhi) [inline src/transformers/models/regnet/modeling_flax_regnet.py:202]: (nit) personally don't think the tensor dim hints are required - I know it's in the PyTorch code but don't think it's super beneficial here!

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+    """
+    Squeeze and Excitation layer (SE) proposed in [Squeeze-and-Excitation Networks](https://arxiv.org/abs/1709.01507).
+    """
+
+    in_channels: int
+    reduced_channels: int
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.pooler = partial(nn.avg_pool, padding=((0, 0), (0, 0)))
```
Comment (sanchit-gandhi) [inline src/transformers/models/regnet/modeling_flax_regnet.py:222]: Nice!

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+            FlaxRegNetConvLayer(
+                self.out_channels,
+                kernel_size=1,
+                activation=None,
+                dtype=self.dtype,
+                name="3",
+            ),
+        ]
+
+    def __call__(self, hidden_state: jnp.ndarray, deterministic: bool = True) -> jnp.ndarray:
```
Comment (sanchit-gandhi) [inline src/transformers/models/regnet/modeling_flax_regnet.py:358]: Deterministic arg unused

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+            # downsampling is done in the first layer with stride of 2
+            layer(
+                self.config,
+                self.in_channels,
+                self.out_channels,
+                stride=self.stride,
+                dtype=self.dtype,
+                name="0",
+            ),
+            *[
```
Comment (sanchit-gandhi) [inline src/transformers/models/regnet/modeling_flax_regnet.py:430]: Can we split this like we did for Flax ResNet please!

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 6]
```diff
@@ -0,0 +1,818 @@
...
+                FlaxRegNetStage(
+                    self.config,
+                    in_channels,
+                    out_channels,
+                    depth=depth,
+                    dtype=self.dtype,
+                    name=str(i + 1),
+                )
+                for i, ((in_channels, out_channels), depth) in enumerate(zip(in_out_channels, self.config.depths[1:]))
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/regnet/modeling_flax_regnet.py:501]: Yes please!

File: src/transformers/models/regnet/modeling_flax_regnet.py (hunk 1 above)
Comment (sanchit-gandhi) [reply src/transformers/models/regnet/modeling_flax_regnet.py:704]: Thanks! Should be merged shortly

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 7]
```diff
@@ -0,0 +1,818 @@
...
+
+
+class FlaxRegNetForImageClassificationModule(nn.Module):
+    config: RegNetConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.regnet = FlaxRegNetModule(config=self.config, dtype=self.dtype)
+
+        if self.config.num_labels > 0:
```
Comment (sanchit-gandhi) [inline src/transformers/models/regnet/modeling_flax_regnet.py:739]: This is fine - we won't change `self.config.num_labels` after doing the compile step, so this will become static after that.

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 8]
```diff
@@ -0,0 +1,818 @@
...
+    >>> image_processor = AutoImageProcessor.from_pretrained("facebook/regnet-y-040")
+    >>> model = FlaxRegNetForImageClassification.from_pretrained("facebook/regnet-y-040")
+
+    >>> inputs = image_processor(images=image, return_tensors="np")
+    >>> outputs = model(**inputs)
+    >>> logits = outputs.logits
+
+    >>> # model predicts one of the 1000 ImageNet classes
+    >>> predicted_class_idx = jax.numpy.argmax(logits, axis=-1)
+    >>> print("Predicted class:", model.config.id2label[predicted_class_idx.item()])
```
Comment (sanchit-gandhi) [inline src/transformers/models/regnet/modeling_flax_regnet.py:809]: Nice!

File: src/transformers/modeling_flax_outputs.py [hunk 9]
```diff
@@ -45,6 +45,63 @@ class FlaxBaseModelOutput(ModelOutput):
     attentions: Optional[Tuple[jnp.ndarray]] = None
 
 
+@flax.struct.dataclass
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_outputs.py:48]: These will be superseded by the FlaxResNet PR

File: tests/models/regnet/test_modeling_flax_regnet.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+    from transformers.models.regnet.modeling_flax_regnet import FlaxRegNetForImageClassification, FlaxRegNetModel
+
+if is_vision_available():
+    from PIL import Image
+
+    from transformers import AutoFeatureExtractor
+
+
+class FlaxRegNetModelTester(unittest.TestCase):
```
Comment (sanchit-gandhi) [inline tests/models/regnet/test_modeling_flax_regnet.py:39]: Could we make the same updates to this tester that we did for Flax ResNet? Namely reducing the image size?

File: tests/models/regnet/test_modeling_flax_regnet.py [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+@require_flax
+class FlaxRegNetModelIntegrationTest(unittest.TestCase):
+    @cached_property
+    def default_feature_extractor(self):
+        return AutoFeatureExtractor.from_pretrained("facebook/regnet-y-040") if is_vision_available() else None
+
+    @slow
+    def test_inference_image_classification_head(self):
+        model = FlaxRegNetForImageClassification.from_pretrained("Shubhamai/regnet-y-040")
```
Comment (sanchit-gandhi) [inline tests/models/regnet/test_modeling_flax_regnet.py:226]: And update this to the official model

**Review by sanchit-gandhi — COMMENTED** [review]
Very nice PR @Shubhamai! Just some small code clean-up comments - the JAX/Flax logic looks sound! In general, we can try and use `# Copied from` statements in every place we copy code from Flax ResNet one-for-one.

File: tests/models/regnet/test_modeling_flax_regnet.py (hunk 11 above)
Comment (Shubhamai) [reply tests/models/regnet/test_modeling_flax_regnet.py]: Updated

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 12]
```diff
@@ -0,0 +1,818 @@
...
+        num_channels = pixel_values.shape[-1]
+        if num_channels != self.config.num_channels:
+            raise ValueError(
+                "Make sure that the channel dimension of the pixel values match with the one set in the configuration."
+            )
+        hidden_state = self.embedder(pixel_values, deterministic=deterministic)
+        return hidden_state
+
+
+class FlaxRegNetShortCut(nn.Module):
```
[the remark this answers is not in the document]
Comment (Shubhamai) [reply src/transformers/models/regnet/modeling_flax_regnet.py:155]: Agree, I have added it wherever possible in the recent commit

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 13]
```diff
@@ -0,0 +1,818 @@
...
+        if output_hidden_states:
+            hidden_states = hidden_states + (hidden_state.transpose(0, 3, 1, 2),)
+
+        if not return_dict:
+            return tuple(v for v in [hidden_state, hidden_states] if v is not None)
+
+        return FlaxBaseModelOutputWithNoAttention(last_hidden_state=hidden_state, hidden_states=hidden_states)
+
+
+class FlaxRegNetPreTrainedModel(FlaxPreTrainedModel):
```
[the remark this answers is not in the document]
Comment (Shubhamai) [reply src/transformers/models/regnet/modeling_flax_regnet.py:553]: Yes! I have added `# Copied ...` in the recent commit

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 14]
```diff
@@ -0,0 +1,818 @@
...
+        outputs = self.regnet(
+            pixel_values,
+            deterministic=deterministic,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        pooled_output = outputs.pooler_output if return_dict else outputs[1]
+
+        logits = self.classifier(pooled_output[:, :, 0, 0])
```
[the remark this answers is not in the document]
Comment (Shubhamai) [reply src/transformers/models/regnet/modeling_flax_regnet.py:766]: No specific reason, since we are using `# Copied from ...` from resnet, this needs to be kept same.

**Comment by Shubhamai** [comment]
@sanchit-gandhi All the requested changes have been made and looks ready for next iteration of review, thanks a lot for your time.

File: src/transformers/models/regnet/modeling_flax_regnet.py [hunk 15]
```diff
@@ -0,0 +1,818 @@
...
+        if not return_dict:
+            return tuple(v for v in [hidden_state, hidden_states] if v is not None)
+
+        return FlaxBaseModelOutputWithNoAttention(
+            last_hidden_state=hidden_state,
+            hidden_states=hidden_states,
+        )
+
+
+# Copied from transformers.models.resnet.modeling_flax_resnet.FlaxResNetPreTrainedModel with ResNet->RegNet,resnet->regnet,RESNET->REGNET
```
Comment (sanchit-gandhi) [inline src/transformers/models/regnet/modeling_flax_regnet.py:552]: Nice!

**Review by sanchit-gandhi — APPROVED** [review]
Very nice @Shubhamai - thanks for incorporating all of the comments from the Flax ResNet PR and the previous Flax RegNet review! 

Looks ready for a final review!

**Review by sgugger — APPROVED** [review]
Thanks for adding this new model! Could you quickly rebase your branch on main? it just fix the issues with the CI.
